pennylane/beta/tapes/qnode.py

```diff
@@ -0,0 +1,528 @@
+        """Validates whether a particular device and QuantumTape interface
+        supports the parameter-shift differentiation method, and returns
+        the correct tape.
+
+        Args:
+            device (.Device): PennyLane device
+            interface (str): name of the requested interface
+
+        Returns:
+            tuple[.QuantumTape, str, str]: tuple containing the compatible
+            QuantumTape, the interface to apply, and the method argument
+            to pass to the ``QuantumTape.jacobian`` method.
+
+        Raises:
+            QuantumFunctionError: if the device model does not have a corresponding
+            parameter-shift rule.
+        """
+        # determine if the device provides its own jacobian method
+        model = device.capabilities().get("model", None)
+
+        if model == "qubit":
+            return QubitParamShiftTape, interface, "analytic"
+
+        if model == "cv":
+            warnings.warn(
+                "CV parameter-shift rule not yet implemented! Falling back "
+                "to finite-differences."
+            )
+
+            return QuantumTape, interface, "numeric"
+
+        raise QuantumFunctionError(
+            f"Device {device.short_name} has model {model} "
+            "that does not support the parameter-shift rule."
+        )
+
+    def construct(self, args, kwargs):
+        """Call the quantum function with a tape context,
+        ensuring the operations get queued."""
+        self.qtape = self._tape()
+
+        # apply the interface (if any)
+        if self.interface is not None:
+            self.INTERFACE_MAP[self.interface](self)
+
+        with self.qtape:
+            measurements = self.func(*args, **kwargs)
+
+        if not isinstance(measurements, Sequence):
+            measurements = (measurements,)
+
+        if not all(isinstance(m, MeasurementProcess) for m in measurements):
+            raise QuantumFunctionError(
+                "A quantum function must return either a single measured observable "
+                "or a nonempty sequence of measured observables."
+            )
+
+        if not all(ret == m for ret, m in zip(measurements, self.qtape.measurements)):
+            raise QuantumFunctionError(
+                "All measurements must be returned in the order they are measured."
+            )
+
+        # provide the jacobian options
+        self.qtape.jacobian_options = self.diff_options
+
+        stop_at = self.device.operations
+
+        if isinstance(self.qtape, QubitParamShiftTape):
+            # controlled rotations aren't supported by the parameter-shift rule
+            stop_at = set(self.device.operations) - {"CRX", "CRZ", "CRY", "CRot"}
+
+        # expand out the tape, if any operations are not supported on the device
+        if not {op.name for op in self.qtape.operations}.issubset(stop_at):
+            self.qtape = self.qtape.expand(depth=self.max_expansion, stop_at=stop_at)

```

Comment:
@trbromley I temporarily added this. It expands out the tape to remove CRm gates if the tape is a parameter shift one.